File: CalSnap/Core/Utilities/AnalyticsTypes.swift
import Foundation

enum AnalyticsDateRange: Hashable, Sendable {
    case days(Int)
    case custom(start: Date, end: Date)

    static let maxCustomSpanDays = 365

    func resolvedEnd(reference: Date, calendar: Calendar = .current) -> Date {
        switch self {
        case .days:
            return calendar.startOfDay(for: reference)
        case .custom(_, let end):
            return calendar.startOfDay(for: min(end, reference))
        }
    }

    func resolvedStart(reference: Date, calendar: Calendar = .current) -> Date {
        let end = resolvedEnd(reference: reference, calendar: calendar)
        switch self {
        case .days(let count):
            return calendar.date(byAdding: .day, value: -(count - 1), to: end) ?? end
        case .custom(let start, _):
            let normalizedStart = calendar.startOfDay(for: start)
            let normalizedEnd = end
            if normalizedStart <= normalizedEnd {
                return normalizedStart
            }
            return normalizedEnd
        }
    }

    var displayLabel: String {
        switch self {
        case .days(7): return String(localized: "model.analytics.timeframe.7d")
        case .days(30): return String(localized: "model.analytics.timeframe.30d")
        case .days(90): return String(localized: "model.analytics.timeframe.90d")
        case .days(let n): return String(format: String(localized: "model.analytics.timeframe.daysFormat"), n)
        case .custom: return String(localized: "model.analytics.timeframe.custom")
        }
    }
}

enum AnalyticsTimeframePreset: String, CaseIterable, Identifiable {
    case days7 = "7D"
    case days30 = "30D"
    case days90 = "90D"
    case custom = "Custom"

    var id: String { rawValue }

    var dateRange: AnalyticsDateRange {
        switch self {
        case .days7: return .days(7)
        case .days30: return .days(30)
        case .days90: return .days(90)
        case .custom: return .days(7)
        }
    }
}

struct DailyNutritionSummary: Sendable, Identifiable, Hashable {
    let date: Date
    let calories: Int
    let proteinG: Double
    let carbsG: Double
    let fatG: Double
    let fiberG: Double

    var id: Date { date }
}

struct MacroSplit: Sendable, Equatable {
    let proteinPct: Int
    let carbsPct: Int
    let fatPct: Int
}

enum Weekday: Int, CaseIterable, Sendable {
    case sunday = 1
    case monday
    case tuesday
    case wednesday
    case thursday
    case friday
    case saturday

    init?(calendarWeekday: Int) {
        self.init(rawValue: calendarWeekday)
    }

    var shortLabel: String {
        switch self {
        case .sunday: String(localized: "model.analytics.weekday.sun")
        case .monday: String(localized: "model.analytics.weekday.mon")
        case .tuesday: String(localized: "model.analytics.weekday.tue")
        case .wednesday: String(localized: "model.analytics.weekday.wed")
        case .thursday: String(localized: "model.analytics.weekday.thu")
        case .friday: String(localized: "model.analytics.weekday.fri")
        case .saturday: String(localized: "model.analytics.weekday.sat")
        }
    }

    var isWeekend: Bool {
        self == .saturday || self == .sunday
    }
}

enum TimeOfDayBucket: String, CaseIterable, Sendable {
    case morning
    case midday
    case evening
    case night

    var displayLabel: String {
        switch self {
        case .morning: String(localized: "model.analytics.timeOfDay.morning")
        case .midday: String(localized: "model.analytics.timeOfDay.midday")
        case .evening: String(localized: "model.analytics.timeOfDay.evening")
        case .night: String(localized: "model.analytics.timeOfDay.night")
        }
    }

    static func bucket(for hour: Int) -> TimeOfDayBucket {
        switch hour {
        case 5..<11: return .morning
        case 11..<15: return .midday
        case 15..<21: return .evening
        default: return .night
        }
    }
}

struct TopFoodEntry: Sendable, Identifiable, Equatable {
    let name: String
    let count: Int
    let avgCalories: Int

    var id: String { name }
}

struct AnalyticsInsightPayload: Sendable {
    let timeframeLabel: String
    let loggedDayCount: Int
    let averageDailyCalories: Int
    let calorieTarget: Int
    let adherencePercent: Double
    let actualMacroSplit: MacroSplit
    let targetMacroSplit: MacroSplit
    let averageDailyFiberG: Double
    let fiberTargetG: Double
    let weekendAverageCalories: Int?
    let weekdayAverageCalories: Int?
    let topFoods: [TopFoodEntry]
    let weightChangeKg: Double?
}

File: calsnap-web/lib/analytics/analytics-types.ts
import { daysBetween, localDayKey, startOfLocalDay } from '@/lib/dashboard/date-window';

const ANALYTICS_MAX_CUSTOM_SPAN_DAYS = 365;

export const ANALYTICS_MIN_INSIGHT_LOGGED_DAYS = 3;

export type AnalyticsDateRange =
  | { kind: 'days'; count: number }
  | { kind: 'custom'; start: Date; end: Date };

export const AnalyticsDateRange = {
  days(count: number): AnalyticsDateRange {
    return { kind: 'days', count };
  },

  custom(start: Date, end: Date): AnalyticsDateRange {
    return { kind: 'custom', start, end };
  },

  resolvedEnd(range: AnalyticsDateRange, reference: Date = new Date()): Date {
    const today = startOfLocalDay(reference);
    if (range.kind === 'days') {
      return today;
    }
    const endDay = startOfLocalDay(range.end);
    return endDay.getTime() <= today.getTime() ? endDay : today;
  },

  resolvedStart(range: AnalyticsDateRange, reference: Date = new Date()): Date {
    const end = AnalyticsDateRange.resolvedEnd(range, reference);
    if (range.kind === 'days') {
      const start = new Date(end);
      start.setDate(start.getDate() - (range.count - 1));
      return startOfLocalDay(start);
    }
    const normalizedStart = startOfLocalDay(range.start);
    const normalizedEnd = end;
    if (normalizedStart.getTime() <= normalizedEnd.getTime()) {
      return normalizedStart;
    }
    return normalizedEnd;
  },

};

export type AnalyticsTimeframePreset = '7D' | '30D' | '90D' | 'custom';

export const ANALYTICS_TIMEFRAME_PRESETS: AnalyticsTimeframePreset[] = [
  '7D',
  '30D',
  '90D',
  'custom',
];

export function presetToDateRange(preset: AnalyticsTimeframePreset): AnalyticsDateRange {
  switch (preset) {
    case '7D':
      return AnalyticsDateRange.days(7);
    case '30D':
      return AnalyticsDateRange.days(30);
    case '90D':
      return AnalyticsDateRange.days(90);
    case 'custom':
      return AnalyticsDateRange.days(7);
  }
}

export function analyticsRangeKey(
  range: AnalyticsDateRange,
  reference: Date = new Date(),
): string {
  if (range.kind === 'days') {
    if (range.count === 7) return '7d';
    if (range.count === 30) return '30d';
    if (range.count === 90) return '90d';
    return `${range.count}d`;
  }
  const start = AnalyticsDateRange.resolvedStart(range, reference);
  const end = AnalyticsDateRange.resolvedEnd(range, reference);
  return `custom:${localDayKey(start)}:${localDayKey(end)}`;
}

export function normalizeCustomRange(
  start: Date,
  end: Date,
  reference: Date = new Date(),
): AnalyticsDateRange {
  let startDay = startOfLocalDay(start);
  let endDay = startOfLocalDay(end);
  const today = startOfLocalDay(reference);

  if (endDay.getTime() > today.getTime()) {
    endDay = today;
  }

  if (startDay.getTime() > endDay.getTime()) {
    [startDay, endDay] = [endDay, startDay];
  }

  const spanDays = daysBetween(startDay, endDay) + 1;
  if (spanDays > ANALYTICS_MAX_CUSTOM_SPAN_DAYS) {
    const trimmedStart = new Date(endDay);
    trimmedStart.setDate(trimmedStart.getDate() - (ANALYTICS_MAX_CUSTOM_SPAN_DAYS - 1));
    startDay = startOfLocalDay(trimmedStart);
  }

  return AnalyticsDateRange.custom(startDay, endDay);
}

export function analyticsWindow(
  range: AnalyticsDateRange,
  reference: Date = new Date(),
): { start: Date; end: Date } {
  let start = AnalyticsDateRange.resolvedStart(range, reference);
  let end = AnalyticsDateRange.resolvedEnd(range, reference);
  const today = startOfLocalDay(reference);

  if (end.getTime() >= today.getTime()) {
    end = new Date(end);
    end.setDate(end.getDate() - 1);
    if (range.kind === 'days') {
      const newStart = new Date(end);
      newStart.setDate(newStart.getDate() - (range.count - 1));
      start = startOfLocalDay(newStart);
    }
  }

  return { start, end };
}

export interface DailyNutritionSummary {
  date: Date;
  calories: number;
  proteinG: number;
  carbsG: number;
  fatG: number;
  saturatedFatG: number;
  unsaturatedFatG: number;
  fiberG: number;
}

export interface TopFoodEntry {
  name: string;
  count: number;
  avgCalories: number;
}
